### PR TITLE
remove mono repo dependency from create-plugin

### PR DESCRIPTION
**Pull request by feiming**
Remove mono repo dependency from create-plugin
Add parameters to support private npm registry

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: packages/cli/src/lib/tasks.ts [hunk 1]
```diff
@@ -73,6 +74,10 @@ export async function templatingTask(
     throw new Error(`Failed to read template directory: ${error.message}`);
   });
 
+  const isMonoRepo = (await fs.pathExists(paths.resolveTargetRoot('plugins')))
```
Comment (Rugvip) [inline packages/cli/src/lib/tasks.ts:77]: This isn't actually true for apps created with `create-app`, even though it's a monorepo. The simplest check for commands that run inside packages is `paths.targetRoot === paths.targetDir`, but since this is supposed to be run at the root I think checking for `paths.resolveTargetRoot(lerna.json)` is the most sensible option.

File: packages/cli/templates/default-plugin/package.json.hbs [hunk 2]
```diff
@@ -1,11 +1,12 @@
 {
-  "name": "@backstage/plugin-{{id}}",
+  "name": "{{#if scopeName}}{{scopeName}}/{{/if}}/plugin-{{id}}",
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/package.json.hbs:2]: probably just template the entire thing as `{{ packageName }}` tbh, and put this together in the calling code.

File: packages/cli/templates/default-plugin/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "extends": "@backstage/cli/config/tsconfig.json",
+  "include": [
+      "src",
+      "dev",
+      "migrations"
+  ],
+  "exclude": ["node_modules"],
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/tsconfig.json:8]: No need to exclude node_modules

File: packages/cli/templates/default-plugin/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  "compilerOptions": {
+    "outDir": "dist-types",
+    "skipLibCheck": true
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/tsconfig.json:11]: Could skip this as it won't be needed after #2317

File: package.json [hunk 5]
```diff
@@ -20,7 +20,8 @@
     "docgen": "lerna run docgen",
     "docker-build:app": "yarn workspace example-app build && docker build . -t spotify/backstage",
     "docker-build": "yarn tsc && yarn workspace example-backend build-image",
-    "create-plugin": "backstage-cli create-plugin",
+    "create-plugin": "backstage-cli create-plugin --scope internal",
```
Comment (Rugvip) [inline package.json:23]: In this repo we'd just have a single `create-plugin`, with `--scope backstage --no-private`

File: packages/cli/src/lib/tasks.ts [hunk 6]
```diff
@@ -73,6 +74,10 @@ export async function templatingTask(
...
+    ? true
```
Comment (Rugvip) [inline packages/cli/src/lib/tasks.ts:78]: This ternary is redundant :grin:

File: packages/cli/templates/default-plugin/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+  "extends": "@backstage/cli/config/tsconfig.json",
+  "include": [
+      "src",
+      "dev",
+      "migrations"
+  ],
+  "exclude": ["node_modules"],
+  "compilerOptions": {
+    "outDir": "dist-types",
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/tsconfig.json:10]: Best set `"rootDir": "."` too, just in case

**Review by Rugvip — COMMENTED** [review]
Thanks, getting there! :grin:

File: package.json (hunk 5 above)
Comment (feiming) [reply package.json:23]: ok

File: packages/cli/templates/default-plugin/tsconfig.json (hunk 3 above)
Comment (feiming) [reply packages/cli/templates/default-plugin/tsconfig.json:8]: ok

File: packages/cli/templates/default-plugin/tsconfig.json (hunk 4 above)
Comment (feiming) [reply packages/cli/templates/default-plugin/tsconfig.json:11]: ok

File: packages/cli/templates/default-plugin/tsconfig.json (hunk 7 above)
Comment (feiming) [reply packages/cli/templates/default-plugin/tsconfig.json:10]: ok

File: packages/cli/src/lib/tasks.ts (hunk 1 above)
Comment (feiming) [reply packages/cli/src/lib/tasks.ts:77]: I tried `paths.targetRoot === paths.targetDir` on scaffold plugin outside of monorepo, it doesn't work.

I'll use `paths.resolveTargetRoot(lerna.json)` to detect monorepo.

File: packages/cli/src/lib/tasks.ts (hunk 6 above)
Comment (feiming) [reply packages/cli/src/lib/tasks.ts:78]: yup.

File: packages/cli/templates/default-plugin/package.json.hbs [hunk 8]
```diff
@@ -1,11 +1,12 @@
 {
-  "name": "@backstage/plugin-{{id}}",
+  "name": "{{name}}",
   "version": "{{version}}",
   "main": "src/index.ts",
   "types": "src/index.ts",
   "license": "Apache-2.0",
-  "private": true,
+  "private": {{privatePackage}},
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/package.json.hbs:7]: These additions are breaking the `plugin:diff` command, as it execute the templates and then compare output. We need to add logic to the `diff` part of the cli to read out the config values for each plugin, and then supply them to the templating. That stuff is happening here: https://github.com/spotify/backstage/blob/87699ffe202eff366a8932f7b1926c7c965905f9/packages/cli/src/commands/plugin/diff.ts#L78

Comment (feiming) [reply packages/cli/templates/default-plugin/package.json.hbs:7]: i'm working on it.

Comment (feiming) [reply packages/cli/templates/default-plugin/package.json.hbs:7]: We also can't extract `version` from lerna.json. Plugins will eventually have a different version from backstage.

https://github.com/spotify/backstage/blob/f90d645461bf82f5745c262e042671c6b04cb362/packages/cli/src/commands/plugin/diff.ts#L65

File: packages/cli/src/commands/create-plugin/createPlugin.ts [hunk 9]
```diff
@@ -221,20 +224,28 @@ export default async () => {
...
   const appPackage = paths.resolveTargetRoot('packages/app');
   const templateDir = paths.resolveOwn('templates/default-plugin');
   const tempDir = resolvePath(os.tmpdir(), answers.id);
-  const pluginDir = paths.resolveTargetRoot('plugins', answers.id);
+  const pluginDir = isMonoRepo
+    ? paths.resolveTargetRoot('plugins', answers.id)
+    : paths.resolveTargetRoot(answers.id);
   const ownerIds = parseOwnerIds(answers.owner);
-  const { version } = await fs.readJson(paths.resolveTargetRoot('lerna.json'));
+  const version = backstageVersion;
```
Comment (Rugvip) [inline packages/cli/src/commands/create-plugin/createPlugin.ts:241]: Why was this changed?

**Review by Rugvip — COMMENTED** [review]
Thanks, looks good! Just one thing to iron out

File: packages/cli/src/commands/create-plugin/createPlugin.ts (hunk 9 above)
Comment (freben) [reply packages/cli/src/commands/create-plugin/createPlugin.ts:241]: In another comment they mention `We can't extract version from lerna.json. Plugins will eventually have a different version from backstage. I'll use version from plugin's package.json`, could that be related?

Comment (Rugvip) [reply packages/cli/src/commands/create-plugin/createPlugin.ts:241]: Ah, so reading back I'm guessing this was to make the diff work?

In that case, revert this line to the original, and grab the version from the target `package.json` itself during diff, instead of `lerna.json`

Comment (feiming) [reply packages/cli/src/commands/create-plugin/createPlugin.ts:240]: during diff, it grabs from package.json. 
https://github.com/spotify/backstage/pull/2316/files#diff-8d46217384c7cb4b4188d65bc1341d83R87
 
By using `lerna.js` here, we can't create-plugin outside of monorepo.

**Review by Rugvip — APPROVED** [review]
Awesome, thanks!
